src/raftstore/coprocessor/split_check_observer.rs

```diff
@@ -0,0 +1,664 @@
+            // `before_check` found a split key.
+            Some(keys::data_key(&last_encoded_table_prefix))
+        } else if let Some(current_encoded_table_prefix) =
+            cross_table(&status.first_encoded_table_prefix, key)
+        {
+            Some(keys::data_key(&current_encoded_table_prefix))
+        } else {
+            None
+        }
+    }
+
+    /// If `current_data_key` does not belong to `encoded_table_prefix`.
+    /// it returns the `current_data_key`'s encoded table prefix.
+    fn cross_table(encoded_table_prefix: &[u8], current_data_key: &[u8]) -> Option<Vec<u8>> {
+        if !keys::validate_data_key(current_data_key) {
+            return None;
+        }
+        let current_encoded_key = keys::origin_key(current_data_key);
+        if !current_encoded_key.starts_with(table_codec::TABLE_PREFIX) {
+            return None;
+        }
+
+        let current_encoded_table_prefix =
+            if let Some(key) = extract_encoded_table_prefix(current_encoded_key) {
+                key
+            } else {
+                return None;
+            };
+
+        if encoded_table_prefix != current_encoded_table_prefix {
+            Some(current_encoded_table_prefix.to_vec())
+        } else {
+            None
+        }
+    }
+
+    #[allow(collapsible_if)]
+    fn bound_keys(db: &DB, region: &Region) -> Result<Option<(Vec<u8>, Vec<u8>)>> {
+        let start_key = keys::enc_start_key(region);
+        let end_key = keys::enc_end_key(region);
+        let mut first_key = None;
+        let mut last_key = None;
+
+        let iter_opt = IterOption::new(Some(end_key), false);
+        let mut iter = box_try!(db.new_iterator_cf(CF_WRITE, iter_opt));
+
+        // the first key
+        if iter.seek(start_key.as_slice().into()) {
+            let key = iter.key().to_vec();
+            first_key = Some(key);
+        } // else { No data in this CF }
+
+        // the last key
+        if iter.seek(SeekKey::End) {
+            let key = iter.key().to_vec();
+            last_key = Some(key);
+        } // else { No data in this CF }
+
+        match (first_key, last_key) {
+            (Some(fk), Some(lk)) => Ok(Some((fk, lk))),
+            (None, None) => Ok(None),
+            (first_key, last_key) => Err(box_err!(
+                "invalid bound, first key: {:?}, last key: {:?}",
+                first_key,
+                last_key
+            )),
+        }
+    }
+
+    // Encode a key like `t{i64}` will append some unecessary bytes to the output,
+    // This function extracts the first 10 bytes, they are enough to find out which
+    // table this key belongs to.
+    #[inline]
+    fn extract_encoded_table_prefix(encode_key: &[u8]) -> Option<&[u8]> {
+        const LEN: usize = table_codec::TABLE_PREFIX_KEY_LEN + 1;
+        if encode_key.len() >= LEN {
+            Some(&encode_key[..LEN])
+        } else {
+            None
+        }
+    }
+
+    #[cfg(test)]
+    mod test {
```

Comment:
any test to split meta and table.

Reply:
Already covered at https://github.com/pingcap/tikv/pull/2378/files/a975491b8ea97bf3a9203b5c3ef3fd2bfdf7bfa6#diff-71afbb86221abfe36cab2ba707276b77R643

Reply:
Better to explicitly use prefix `m`